### PR TITLE
Fix ToByteArray memory corruption.

### PR DESCRIPTION
**Pull request by kovacsv**
It should fix the memory corruption in the toByteArray function called from javascript.

**Error**
The previous approach just created a view on the buffer that is owned by the `ON_Write3dmBufferArchive` class. At the end of the `ToByteArray2` function the memory is freed by the owner. So it was the question of luck if it works or not.

**Solution**
Now I create a new buffer for the result, and copy the original buffer content to it. Returning of this value is somewhat cleaner than the original solution. To be honest it introduces another problem: Who will free this memory up? I don't have a solution for that right now (maybe another function that should be called when the buffer is not needed anymore), but at least now it works as expected.

**Fixes**
- Sure:
  - https://discourse.mcneel.com/t/rhino3dm-js-fails-to-write-a-single-cube-to-a-3dm-file/123467/2
- Probably:
  - https://discourse.mcneel.com/t/rhino3dm-js-fails-to-write-a-single-cube-to-a-3dm-file/123467/2
  - https://discourse.mcneel.com/t/rhino3dm-threejs-geometry-import-failing/118635
  - toByteArray produces invalid 3dm file in some cases #353

### CODE REVIEW

**Comment by sbaer** [comment]
I’m concerned about the memory leak that this causes. Could you help me understand a bit more why the current approach doesn’t work? I don’t understand the comments on ownership and emscripten is a bit light on this topic.

Thanks

**Comment by kovacsv** [comment]
This is how I understand the old situation:
- There is a buffer called `buffer` which contains the file content. This is just a pointer to a member of the `ON_Read3dmBufferArchive` class. When the `ON_Read3dmBufferArchive` goes out of scope this buffer is released.
- There is a `Uint8Array` that points to this buffer. In javascript the `Uint8Array` is not a real buffer, just a view to another buffer. In the old approach the created `Uint8Array` just pointed to the current heap section where the `buffer` is stored.
- At the end of the `ToByteArray2` function the `buffer` is released by the destructor of `ON_Read3dmBufferArchive`.

When you delete an pointer usually the memory don't be erased and with some luck you can access it right after deletion. This is why it worked in most of the cases. But some cases the memory area is already reused and this is why it failed sometimes. It seems that some other code wrote in the released memory area, so it messed up the `Uint8Array` that is later used in javascript.

So to summarize it: the function returned a buffer that was released right after the function is finished. In some cases the released buffer was already reused by some other code, and this caused the issue.

**Comment by sbaer** [comment]
That was never the intent of the original code and is why I created an instance of a new Uint8Array that gets filled with a copy of the contents of the ON_3dmBufferArchive buffer.

Maybe my understanding of what Uint8Array is in the context of emscripten is where I have problems. My intent was to create a new Uint8Array of a defined size / content and pass that back to javascript. If that Uint8Array content is still pointing at wasm memory, then I can see where problems would occur.

**Comment by kovacsv** [comment]
So here is the relevant code segment from the original code:
```cpp
emscripten::val rc = Uint8Array.new_(emscripten::val::module_property("HEAPU8")["buffer"], size_t(buffer), length);
```

It calls the three-parameter constructor of Uint8Array (buffer, byteOffset, length). According to the [specification](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/Uint8Array) the three-parameter constructor creates only a typed array **view** that views the specified buffer, and doesn't allocate a new buffer.

The next loop just writes the same values in the same buffer, so it doesn't copy anything.
```cpp
for (size_t i = 0; i < length; i++)
{
  rc.set(i, buffer[i]);
}
```

So as I understand no allocation or copy is happening here, we just create a view to the buffer, and that buffer is freed at the end of the function.

I'm not an expert in Emscripten, but kripken is (who wrote the whole stuff), and he mentions the same here:
https://github.com/emscripten-core/emscripten/issues/13065

**Comment by sbaer** [comment]
I see, thanks. In that case we might as well new up the ON_3dmBufferArchive and directly map in the the archive's buffer which eliminates the need for a secondary new/memcpy. We can keep a list of these archives and provide a function to delete them.

```
ON_Write3dmBufferArchive* trackedArchive = new ON_Write3dmBufferArchive(0, 0, options->VersionForWriting(), ON::Version());
// save pointer to trackedArchive that can be deleted in a different function
... write to trackedArchive ...
const unsigned char* buffer = (const unsigned char*)trackedArchive->Buffer();
size_t length = trackedArchive->SizeOfArchive();

return emscripten::val(emscripten::typed_memory_view(length, buffer));
```

**Comment by kovacsv** [comment]
I think I've found the best solution by combining the old and the new code. We can still create a `Uint8Array` object in C++. The only difference is to use the one-parameter constructor with a typedArray, so the buffer will be copied and maintained by JavaScript. So a copy still happens, but we don't have to deal with freeing up the memory after using it from JavaScript.

The code will look something like this:
```cpp
emscripten::val Uint8Array = emscripten::val::global("Uint8Array");
emscripten::val rc = Uint8Array.new_(emscripten::typed_memory_view(length, buffer));
return rc;
```

I've updated this pull request with the final version, please take a look!

**Comment by sbaer** [comment]
Oh, that looks like exactly what we need! Thanks for all of the help on this topic.

**Comment by kovacsv** [comment]
Thanks, @sbaer!
